src/role_management.c

```diff
@@ -0,0 +1,685 @@
+		goto finish_stop;
+	}
+	leader_addr = raft_malloc(strlen(borrowed_addr) + 1);
+	if (leader_addr == NULL) {
+		goto finish_stop;
+	}
+	memcpy(leader_addr, borrowed_addr, strlen(borrowed_addr) + 1);
+
+	/* Select a non-voter to transfer to -- the logic is similar to adjustClusterCb. */
+	for (i = 0; i < n_cluster; i += 1) {
+		if (cluster[i].online && cluster[i].role == DQLITE_VOTER && cluster[i].id != node->raft.id) {
+			addDomain(cluster[i].failure_domain, &voter_compare);
+		}
+	}
+	qsort_r(cluster, n_cluster, sizeof *cluster, compareNodesForPromotion, &voter_compare);
+	target_id = 0;
+	for (i = 0; i < n_cluster; i += 1) {
+		if (cluster[i].online && cluster[i].role != DQLITE_VOTER && cluster[i].id != node->raft.id) {
+			target_id = cluster[i].id;
+			break;
+		}
+	}
+	/* If no transfer candidates found, give up. */
+	if (target_id == 0) {
+		goto err_after_alloc_leader_addr;
+	}
+
+	/* Submit the handover work. */
+	data = raft_malloc(sizeof *data);
+	if (data == NULL) {
+		goto err_after_alloc_leader_addr;
+	}
+	data->node = node;
+	data->target_id = target_id;
+	data->leader_addr = leader_addr;
+	data->leader_id = leader_id;
+	work = raft_malloc(sizeof *work);
+	if (work == NULL) {
+		goto err_after_alloc_data;
+	}
+	work->data = data;
+	rv = uv_queue_work(&node->loop, work, handoverVoterWorkCb, handoverVoterAfterWorkCb);
+	if (rv != 0) {
+		goto err_after_alloc_work;
+	}
+	return;
+
+err_after_alloc_work:
+	raft_free(work);
+err_after_alloc_data:
+	raft_free(data);
+err_after_alloc_leader_addr:
+	raft_free(leader_addr);
+finish_stop:
+	dqlite__finish_stop(node);
+}
+
+static void handoverTransferCb(struct raft_transfer *req)
+{
+	struct dqlite_node *d = req->data;
+	raft_free(req);
+	pollCluster(d, handoverVoterCb);
+}
+
+void RoleManagementAdjust(struct dqlite_node *d)
+{
+
+	/* Only the leader can assign roles. */
+	if (raft_state(&d->raft) != RAFT_LEADER) {
+		return;
+	}
+	/* If a series of role adjustments is already in progress, don't kick off
+	 * another one. */
+	if (!QUEUE__IS_EMPTY(&d->changes)) {
+		return;
+	}
+	pollCluster(d, adjustClusterCb);
+}
+
+void RoleManagementHandover(struct dqlite_node *d)
+{
+	struct raft_transfer *req;
+	int rv;
+
+	req = raft_malloc(sizeof *req);
+	if (req == NULL) {
+		goto err;
+	}
+	req->data = d;
+	/* We try the leadership transfer unconditionally -- Raft will tell us if
+	 * we're not the leader. */
+	rv = raft_transfer(&d->raft, req, 0, handoverTransferCb);
+	if (rv == RAFT_NOTLEADER) {
+		raft_free(req);
+		pollCluster(d, handoverVoterCb);
+		return;
+	} else if (rv != 0) {
+		raft_free(req);
+		goto err;
+	}
+	return;
+
+err:
+	dqlite__finish_stop(d);
+}
+
+void RoleManagementDrain(struct dqlite_node *d)
+{
+	queue *root = &d->changes;
+	queue *head = QUEUE__NEXT(root);
+	struct change_record *rec;
+	QUEUE__INIT(&d->changes);
+	while (head != root) {
+		rec = QUEUE__DATA(head, struct change_record, queue);
+		head = QUEUE__HEAD(&rec->queue);
+		raft_free(rec);
+	}
```

Comment:
If I understand the goal of the code correctly, a more idiomatic (and probably more readable) way to do this would be:

```C
	while (!QUEUE__IS_EMPTY(&d->changes)) {
		struct change_record *rec;
		queue *head;
		head = QUEUE__HEAD(&d->changes);
		rec = QUEUE__DATA(head, struct change_record, queue);
		QUEUE__REMOVE(head);
		raft_free(rec);
	}
```

There are a few examples of this in the dqlite and raft code base.